app/services/workflow_executions/preparation_service.rb

```diff
@@ -0,0 +1,120 @@
+# frozen_string_literal: true
+
+require 'csv'
+require 'tempfile'
+
+module WorkflowExecutions
+  # Service used to Prepare a WorkflowExecution
+  class PreparationService < BaseService
+    def initialize(workflow_execution, user = nil, params = {})
+      super(user, params)
+      @workflow_execution = workflow_execution
+      @samplesheet_headers = %w[sample fastq_1 fastq_2]
+      @samplesheet_rows = []
+    end
+
+    def execute # rubocop:disable Metrics/AbcSize,Metrics/MethodLength
+      # confirm params/permissions
+      # build workflow execution run directory
+      generate_run_dir
+
+      # does it make sense to rename the SamplesWorkflowExecution model to something better
+      # e.g. SampleWorkflowExecutionParams
+      @workflow_execution.samples_workflow_executions.each do |sample_workflow_execution|
+        attachments = parse_attachments_from_samplesheet(sample_workflow_execution.samplesheet_params)
+        samplesheet_params = sample_workflow_execution.samplesheet_params
+
+        attachments.each do |key, attachment|
+          samplesheet_params[key] = copy_attachment_to_run_dir(attachment, sample_workflow_execution)
+        end
+
+        @samplesheet_rows << samplesheet_params
+      end
+
+      # persist samplesheet in run dir
+      @workflow_execution.inputs.attach(io: samplesheet_file, key: generate_input_key('samplesheet.csv'),
+                                        filename: 'samplesheet.csv')
+
+      @workflow_execution.workflow_params = @workflow_execution.workflow_params.merge(
+        {
+          '--input': generate_input_key('samplesheet.csv'),
+          '--outdir': generate_input_key('output')
+        }
+      )
+
+      # mark workflow execution as prepared
+      @workflow_execution.state = 'prepared'
+
+      @workflow_execution.save
+    end
+
+    private
+
+    def generate_run_dir
+      @run_dir = ActiveStorage::Blob.generate_unique_secure_token
+    end
+
+    def parse_attachments_from_samplesheet(samplesheet)
+      attachments = {}
+      # loop through samplesheet_params to fetch attachments
+      # probably stored as `gid://irida/Attachment/1234`
+      samplesheet.each do |key, value|
+        gid = GlobalID.parse(value)
+        next unless gid && gid.model_class == Attachment
+
+        attachments[key] = GlobalID.find(gid)
+      end
+      attachments
+    end
+
+    def generate_input_key(filename, prefix = '')
+      format('%<run_dir>s/%<prefix>s%<filename>s', run_dir: @run_dir, filename:, prefix:)
+    end
+
+    def copy_attachment_to_run_dir(attachment, attachable) # rubocop:disable Metrics/AbcSize
```

Comment:
Line 75-88 only uses `attachment` and not `attachable`, it could be refactored to make it more readable and not trigger the Metrics/AbcSize rule.

I'd say lines 75-88 could be it's own function, or if you want key generation to exist in `copy_attachment_to_run_dir` then make a function for lines 79-88 called `generate_attachment_blob(key, attachment)` that returns `blob`.

Reply:
Fixed in 6116512b